### ISSUE TITLE
Incorrect order of the events in process pending pointer capture section

### ISSUE THREAD

**Issue by NavidZ** [w3c/pointerevents#39]
@RByers, @mustaqahmed, @jacobrossi 

Looking at this section:

https://w3c.github.io/pointerevents/#process-pending-pointer-capture
- It seems weird that firing pointerleave/out happens after firing gotpointercapture. As the other places in the spec mentioned, if a pointer is captured by some element no other element should get those boundary transition events. The same thing happens for pointerenter/over.
- Also the conditions for when to send the out/leave/over/enter events only covers when the pointer is being captured from an un-captured state or vice versa. Meaning that at least one of the pending pointer capture target or pointer capture target should be clear for those transition events to be fired. I can imagine of some cases that this doesn't sound right when the element which has the capture set another element as the pointer capture.

This example shows both of the cases above:
https://output.jsbin.com/riyotu
1. Move the mouse onto the green div and press and keep the left key down.
2. Move the mouse a little bit in the green box (until you get more than ~10 pointermove) which then the pointer capture will be set to the blue div.
3. Move the mouse a little more until the blue div gets at least one pointer move.

Here is the list of events Edge prints:

1 green received pointermove 1
2 green received pointerover 1
3 green received pointerenter 1
4 green received pointerdown 1
======= Set Pointer Capture ======= 
5 green received gotpointercapture 1
6 green received pointermove 1
7 green received pointermove 1
8 green received pointermove 1
9 green received pointermove 1
10 green received pointermove 1
11 green received pointermove 1
12 green received pointermove 1
13 green received pointermove 1
14 green received pointermove 1
15 green received pointermove 1
16 green received pointermove 1
17 green received pointermove 1
======= Re-set Pointer Capture ======= 
18 green received lostpointercapture 1
19 blue received gotpointercapture 1
20 green received pointerout 1
21 green received pointerleave 1
22 blue received pointermove 1

Note that Edge does send pointerout/leave in steps 20 and 21 while neither of pending pointer capture target or pointer capture target were clear (i.e. pointer capture target was green div and pending pointer capture target was blue div) which spec doesn't cover in the conditions. Also I expected events in steps 20 and 21 happens at least before step 19 as per other parts of the spec. Further it feels more natural if they happen even before 18 but I don't have any arguments for this one.

p.s.: Note that step 1 also seems incorrect as pointermove happens before pointerover/enter but that is not part of this issue and it is probably just a bug in Edge.

**Comment by mustaqahmed** [comment]
Thanks Navid. In summary, there are two problems with the spec:
**Problem 1.** Current spec wording misses the firing of out/leave events when the capturing directly "moves" from one element to another. Edge fires the events in a logical manner, not following the spec (from step 5 to end).
**Problem 2'.** Before "blue" receives pointermove in step 22, it should receive pointerover/enter. Spec is silent again as in Problem 1, but Edge follows the spec this time.
**Problem 3.** After "blue" captures the pointer (step 19), "green" shouldn't receive any event. So firing of pointerout/leave to "green" (steps 20, 21) is conceptually wrong even though the spec suggests the same order _if "green" hadn't captured the mouse in step 5_. Edge implementation seems to follow the spec order.

**Comment by mustaqahmed** [comment]
Correction to **Problem 2'** above: This is similar to **Problem 1** but not covered in Navid's repro since mouse never actually entered blue. So my comment about "Edge following the spec" could be wrong here.

**Comment by RByers** [comment]
> Problem 1. Current spec wording misses the firing of out/leave events when the capturing directly "moves" from one element to another. Edge fires the events in a logical manner, not following the spec (from step 5 to end)

I agree we should just fix this in the spec.

> Problem 2'. Before "blue" receives pointermove in step 22, it should receive pointerover/enter. Spec is silent again as in Problem 1, but Edge follows the spec this time.

Why do you say blue should receive pointerover/enter?  Even when captured, I thought an element should get over/enter only when the cursor has crossed into the bounds of the element.

> Problem 3. After "blue" captures the pointer (step 19), "green" shouldn't receive any event. So firing of pointerout/leave to "green" (steps 20, 21) is conceptually wrong even though the spec suggests the same order if "green" hadn't captured the mouse in step 5. Edge implementation seems to follow the spec order.

I agree that seems more logical.  The implications in practice aren't clear to me though - are there scenarios that would work better or worse given this change?

**Comment by mustaqahmed** [comment]
> > Problem 2'. Before "blue" receives pointermove in step 22, it should receive pointerover/enter. Spec is silent again as in Problem 1, but Edge follows the spec this time.
> 
> Why do you say blue should receive pointerover/enter? Even when captured, I thought an element should get over/enter only when the cursor has crossed into the bounds of the element.

You are right, I corrected my mistake after posting it, sorry.

> > Problem 3. After "blue" captures the pointer (step 19), "green" shouldn't receive any event. So firing of pointerout/leave to "green" (steps 20, 21) is conceptually wrong even though the spec suggests the same order if "green" hadn't captured the mouse in step 5. Edge implementation seems to follow the spec order.
> 
> I agree that seems more logical. The implications in practice aren't clear to me though - are there scenarios that would work better or worse given this change?

I don't have any concrete use-case, but in general, if a web API can avoid pitfalls, it should because there is no way to correct copied JS snippets on the web. Here is why I think Problem 3 can be called a pitfall:

While got/lostpointercapture provide the ideal way to argue about "capture state" in JS, web developers (or software engineers in general) seem to rely more on _what works_ than _what is documented_. I fear that "a non-capturing element receiving events" could be taken as an equivalent to "losing capture" by developers because (i) it doesn't fit common sense, and (ii) Navid's repro is really a hidden case---that's why it went unnoticed so far.

**Comment by NavidZ** [comment]
@jacobrossi, @teddink

What do you think about this one? Do you think the proposed way of sending the events makes the event sequence better?

**Comment by RByers** [comment]
On [the call](https://www.w3.org/2016/05/04-pointerevents-minutes.html#item04) it seemed this issue was largely blocked on the big fundamental question of what the right model for transition events during capture is.  Filed w3c/pointerevents#566 to try to get consensus on that.

If we want to move forward in parallel with that, it seems there was agreement that:
1. Firing transition events at one element after a new one has received `gotpointercapture` is a bug
   - Step 20 and 21 should occur before 19 (and possibly before 18) [above](https://github.com/w3c/pointerevents/issues/39#issue-139363428)
2. The spec should indicate that `pointerover` and `pointerout` behave the same as `pointerenter` and `pointerleave` during capture in that they reflect the bounds of the captured element and all it's descendants.

**Comment by NavidZ** [comment]
I believe this can be closed looking at the latest spec as we changed that part of the spec that explains this.
